### ISSUE TITLE
configure NPath Complexity

### ISSUE THREAD

**Issue by bekuno** [cgeo/cgeo#9660]
see https://github.com/cgeo/cgeo/pull/9633#issuecomment-753465274
see https://github.com/cgeo/WhereYouGo/pull/240#issuecomment-753617124

Question to the @cgeo/team 
Currently we have the codacy check for the parameter `NPath Complexity` with the default parameters (error, max value 200).
Sometimes it would make sense to skip this.
We can configure this by changing the severity level and/or the max value in the checkstyle.xml.
E.g.
~~~
        <module name="NPathComplexity">
            <property name="severity" value="info" />
            <property name="max" value="500"/>
        </module>
~~~
Do we want to change this or let it unchanged?

**Comment by Lineflyer** [comment]
While I can not dive into whether it can be avoided or not or how code readability is affected, I at least learned (due to my normal testing job) that its also helpful to keep the path complexity low in order to be able after all to test all different scenarios of a function/method.
Even 200 can lead to execution paths which produces errors which are difficult to debug.

FYI:
Codacy reports 55 reports of NPathComplexity>200 in our code:
https://app.codacy.com/gh/cgeo/cgeo/issues?&filters=W3siaWQiOiJMYW5ndWFnZSIsInZhbHVlcyI6W119LHsiaWQiOiJDYXRlZ29yeSIsInZhbHVlcyI6W119LHsiaWQiOiJMZXZlbCIsInZhbHVlcyI6W119LHsiaWQiOiJQYXR0ZXJuIiwidmFsdWVzIjpbIjU1NDgiXX0seyJpZCI6IkF1dGhvciIsInZhbHVlcyI6W119XQ==

**Comment by moving-bits** [comment]
> Do we want to change this or let it unchanged?

I don't follow the argumentation of a hard limit (may it be 200, 500 or what else) - sometimes there are situations where code is more readable if you put that one extra step into the main function instead of creating yet another sub function only called by that one main function. It all depends on the specific circumstances.

Having said that, I would keep a certain limit, at least at info level, as a reminder for the developers to keep an eye on it, and to refactor if needed.

**Comment by bekuno** [comment]
The default SeverityLevel is `error`, see https://checkstyle.sourceforge.io/config.html#Checker_Properties
Other levels are 
 -    ignore
 -    info
 -    warning

see https://checkstyle.sourceforge.io/property_types.html#SeverityLevel

Details for NPath Complexity: https://checkstyle.sourceforge.io/config_metrics.html#NPathComplexity

**Comment by eddiemuc** [comment]
My 2 cents wrt NPathComplexity rule (and static code analysis in general)

I am a big fan of static code analysis tools! But since they are "just tools" they will always deliver "false positives". I usually like to use them as follows:
* Configure rules and rule boundaries so they cover 90% of all cases, and make these rules mandatory (in concrete case: leave NPathComplexity rule at error level, maybe increase limit a bit but not too much)
* Allow SuppressWarnings annotations to cover the remaining 10% of cases where rule delivers false positive (do not soften the rule just to prevent a few false positives!)
* Each SuppressWarning must be accompanied by a short comment why in this special case this is a false positive
* During code review, this is usually something to pay attention to

In conclusion imho: maybe set from 200 to 300, 400 or 500, but please leave at error level and allow SuppressWarnings with comments

**Comment by eddiemuc** [comment]
Btw: I would love to see some sort of unit test coverage for c:geo. But I suppose this is a different topic...

**Comment by bekuno** [comment]
So lets stay with the current (default) configuration.
And I agree completely with the 2ct from @eddiemuc
